cmd/gateway/hdfs/gateway-hdfs.go

```diff
@@ -384,28 +384,82 @@ func (n *hdfsObjects) listDirFactory() minio.ListDirFunc {
 
 // ListObjects lists all blobs in HDFS bucket filtered by prefix.
 func (n *hdfsObjects) ListObjects(ctx context.Context, bucket, prefix, marker, delimiter string, maxKeys int) (loi minio.ListObjectsInfo, err error) {
-	if _, err := n.clnt.Stat(n.hdfsPathJoin(bucket)); err != nil {
+	fileInfos := make(map[string]os.FileInfo)
+	directoryPath := n.hdfsPathJoin(bucket, prefix) + minio.SlashSeparator
+
+	var directoryFileInfo os.FileInfo
+
+	if directoryFileInfo, err = n.populateDirectoryListing(directoryPath, fileInfos); err != nil {
```

Comment:
getObjectInfo would still cover this prelude listing right? Why are we doing this initially and checking if the dentries are repeating (already preserved) in L404. 

Are there any performance impacts in doing this?

Reply:
> getObjectInfo would still cover this prelude listing right? Why are we doing this initially…

Are you asking why we do `populateDirectoryListing()` here and not just leave it to be within `getObjectInfo`?

`getObjectInfo` now works by simply keying into the `path → FileInfo` map. If a path is missing, it lists its _parent's_ directory (since to get `/dir/file`'s `FileInfo` we need to list `/dir`). So in the case of the "root" path e.g. `/dir` (assuming we did `ls /dir`), it would mean listing `/`, unnecessarily populating the map with `/`'s entries when all we care about is `/dir`. This can be really bad if `/` is huge.

This process makes sense in all other cases because it would be a nested path in a `--recursive` listing.

The original removed code on line 387 performed an explicit network `Stat()` for the target `directoryPath`, which would be one way for us to "bootstrap" the root `FileInfo`, but `FileReader` is already doing this under the hood and helpfully exposes it with the `FileReader.Stat()` accessor, so, knowing that at a minimum we will want to list the first level of a directory anyway, I just do it upfront right here which allows us to preserve `directoryPath`'s `FileInfo` without an extra unnecessary call.

Not sure if I understood you correctly or if that answers your question.

> …and checking if the dentries are repeating (already preserved) in L404.

Sorry I don't follow this part, what do you mean by "checking if the dentries are repeating (already preserved)"? As far as I'm aware, there are no repeating entries.

> Are there any performance impacts in doing this?

For performing the first level's listing up-front? I don't see why. This is doing work that needs to be done either way, and subsequent `populateDirectoryListing()` will only be for things we don't already have.

So in other words:

1. Here, up-front, we populate the cache with the `FileInfo` for `directoryPath` and for its direct children.
2. On `getObjectInfo(directoryPath)`, in cache, no listing necessary
3. On `getObjectInfo(directoryPath.Child)`, in cache, no listing necessary
4. On `getObjectInfo(directoryPath.Child.Child)`, not in cache, so `list(directoryPath.Child)`

Hope that answers your questions.

Reply:
Now i see its purpose to avoid unnecessary listing :+1: